### DISCUSSION TITLE
AeroDyn v15 inputs correction ?

### DISCUSSION THREAD

**Discussion by SimonPuech** [NatLabRockies/ROSCO#101] [Q&A] [answered]
Hi !
I'm trying to use CCBlade with data from AeroDyn v15 blade files.

I saw in ROSCO/ROSCO_toolbox/turbine.py  in the load_blade_info() definition that we offset the chord and twist data from the hub radius (line 512 to 516).
> r = r0 + self.Rhub 
> chord_intfun = interpolate.interp1d(r0,chord0, bounds_error=None, fill_value='extrapolate', kind='zero') 
> chord = chord_intfun(r) 
> theta_intfun = interpolate.interp1d(r0,theta0, bounds_error=None, fill_value='extrapolate', kind='zero') 
> theta = theta_intfun(r)

I can't manage to understand why this correction is needed as i feel like this correction delete data from near hub nodes.

Does someone know what I'm missing ?

Best Regards,
Simon

**Comment by ptrbortolotti** [comment]
Hello Simon,
CCBlade is a blade solver and it does not model the hub. CCBlade defines the blade stations in the radial coordinates, so the first stations will be at location hub_radius. AeroDyn15 defines the blade stations in the blade coordinate system, where the first station will be located at 0 meters. That's why the code going from AD15 to CCBlade shifts everything by the hub radius.
I hope this helps.
Best regards,
Pietro Bortolotti

**Reply by SimonPuech** [reply]
Hello Pietro,

Thanks for the answer ! Unfortunately, I'm still not sure of understanding everything correctly. I understand the radius shift.
As you say, AeroDyn v15 defines the blades station in the blade coordinate system.
For example, for the IEA15 WT with 120m radius and 3m hub radius, AeroDyn v15 radial coordinates (BladeSpann) will go form 0 to 117m which is why we need to add the 3m of hub radius to obtain a radius list going from 3m to 120m in order to work with CCBlade.

However, I don't understand why do we need to shift the chord and twist angle data as the 1st element of the list of chord (or twist) refer to the 0m bladeSpann (AeroDyn v15) or the 1st radius element which is at 3m radius (after the line 512). I feel like lines 513 to 516 shouldn't exist.

To give a small example :
Let's say that our hub radius is 1.5m and that we have a 6.5m radius WT with the following AeroDyn v15 Table :
> BladeSpann = [0, 1, 2, 5 ]
> Twist = [0.15, 0.14, 0.13, 0.12]
> chord = [0.5, 0.4 ,0.3 ,0.1]

The code will give  :
> r = [1.5, 2.5,3.5, 6.5]
> theta = [0.14, 0.13, 0.13, 0.12]
> chord = [0.4, 0.3, 0.3, 0.1]
(as we interpolate with r instead of r0)

While i feel like we should have :
> r = [1.5, 2.5,3.5, 6.5]
> theta = [0.15, 0.14, 0.13, 0.12]
> chord = [0.5, 0.4 ,0.3 ,0.1]

As i mentioned the code seems to delete near hub chord and twist data and to "copy-paste" near tip data. Is it normal?

Best Regards,
Simon Puech

**Comment by nikhar-abbas** [comment, marked as answer]
Hi @SimonPuech,
You are absolutely correct - this is a small bug in the way we are defining the rotor for cc-blade. 

https://github.com/NREL/ROSCO/pull/104 addresses this and will be merged into the develop branch.

Thanks!
Nikhar

**Comment by ptrbortolotti** [comment]
Hello Simon,
thank you for insisting here, it looks like you've unearthed a bug. The ROSCO developers will fix it soon. For the time being, you should be able to move ahead by simply making chord=chord0 and theta=theta0.
Thanks again for working with ROSCO,
Pietro
